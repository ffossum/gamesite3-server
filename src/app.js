const Koa = require('koa');
const Router = require('koa-router');
const jwt = require('koa-jwt');
const apiRouter = require('./api/apiRouter');
const { expireJwtCookie } = require('./api/jwtCookie');

const jwtSecret = process.env.JWT_SECRET;

const app = new Koa();
const router = new Router();

router.use('/api', apiRouter.routes(), apiRouter.allowedMethods());

router.get(
  '/registration',
  jwt({ secret: jwtSecret, cookie: 'jwt', passthrough: true }),
  async (ctx, next) => {
    if (ctx.state.user) {
      ctx.redirect('/');
    }

    await next();
  }
);

router.get('/logout', expireJwtCookie(), ctx => {
  ctx.redirect('back');
});

router.get(
  '*',
  jwt({ secret: jwtSecret, cookie: 'jwt', passthrough: true }),
  ctx => {
    ctx.body = `<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Gamesite 3</title>
  <script src="//localhost:8080/scripts/bundle.js" defer></script>
</head>

<body>
  <div id="root"></div>
</body>
</html>
`;
  }
);

app.use(router.routes(), router.allowedMethods());

module.exports = app;
