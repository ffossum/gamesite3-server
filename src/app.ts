import * as Koa from 'koa';
import * as Router from 'koa-router';
import * as jwt from 'koa-jwt';
const apiRouter = require('./api/apiRouter');
const { expireJwtCookie } = require('./api/jwtCookie');

const jwtSecret = process.env.JWT_SECRET;

const app = new Koa();
const router = new Router();

router.use('/api', apiRouter.routes(), apiRouter.allowedMethods());

router.get(
  /\/(registration|login)\/?$/,
  jwt({ secret: jwtSecret, cookie: 'jwt', passthrough: true }),
  async (ctx: Koa.Context, next: () => Promise<any>) => {
    if (ctx.state.user) {
      ctx.redirect('/');
    }

    await next();
  }
);

router.get('/logout', expireJwtCookie(), (ctx: Koa.Context) => {
  ctx.redirect('back');
});

router.get(
  '*',
  jwt({ secret: jwtSecret, cookie: 'jwt', passthrough: true }),
  (ctx: Koa.Context) => {
    const user = ctx.state.user && {
      id: ctx.state.user.id,
      username: ctx.state.user.username,
    };

    const userTag = user
      ? `<script defer>window.__USER__ = ${JSON.stringify(user)};</script>`
      : '';

    ctx.body = `<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Gamesite 3</title>

  ${userTag}
  <script src="//localhost:8080/scripts/bundle.js" defer></script>
</head>

<body>
  <div id="root"></div>
</body>
</html>
`;
  }
);

app.use(router.routes());
app.use(router.allowedMethods());

export default app;
