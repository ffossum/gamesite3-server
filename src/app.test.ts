import * as request from "supertest";
import app from "./app";

describe("GET / (unauthenticated)", () => {
  test("responds with html", async () => {
    await request(app.callback())
      .get("/")
      .expect(200)
      .expect("Content-Type", /html/)
      .expect(res => {
        expect(res.text).toMatchSnapshot();
      });
  });
});
