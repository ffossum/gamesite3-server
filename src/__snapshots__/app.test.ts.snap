// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET / (unauthenticated) responds with html 1`] = `
"<!doctype html>

<html lang=\\"en\\">
<head>
  <meta charset=\\"utf-8\\">
  <title>Gamesite 3</title>
  <link rel=\\"shortcut icon\\" href=\\"data:image/x-icon;,\\" type=\\"image/x-icon\\" />
  
  <script src=\\"//localhost:8080/scripts/bundle.js\\" defer></script>
</head>

<body>
  <div id=\\"root\\"></div>
</body>
</html>
"
`;
